_sips/sips/2017-10-25-adding-prefix-types.md

```diff
@@ -0,0 +1,163 @@
+---
+layout: sip
+discourse: true
+title: SIP-NN - Adding prefix types
+
+vote-status: pending
+permalink: /sips/:title.html
+---
+
+**By: Oron Port**
+
+## History
+
+| Date          | Version                                  |
+| ------------- | ---------------------------------------- |
+| Oct 25th 2017 | Split prefix types from [SIP33](http://docs.scala-lang.org/sips/priority-based-infix-type-precedence.html), and emphasize motivation |
+
+
+Your feedback is welcome! If you're interested in discussing this proposal, head over to [this](https://contributors.scala-lang.org/t/sip-nn-make-infix-type-alias-precedence-like-expression-operator-precedence/471) Scala Contributors thread and let me know what you think.
+
+---
+
+## Introduction
+Currently scala supports unary prefix operators (`-`, `+`, `~`, `!`) for expressions (e.g., `def unary_-`) and does not support prefix types.  See the Scala specification [Prefix Operations](http://scala-lang.org/files/archive/spec/2.12/06-expressions.html#prefix-operations) section.
+
+**Prefix expression vs. prefix type example**:
+
+```scala
+object PrefixExpression {
+  case class Nummy(expand : String) {
+    def unary_- : Nummy = Nummy(s"-$this")
+    def unary_~ : Nummy = Nummy(s"~$this")
+    def unary_! : Nummy = Nummy(s"!$this")
+    def unary_+ : Nummy = Nummy(s"+$this")
+  }
+  object N extends Nummy("N")
+  val n1 = -N
+  val n2 = ~N
+  val n3 = !N
+  val n4 = +N
+}
+object NonExistingPrefixTypes {
+  trait unary_-[A]
+  trait unary_~[A]
+  trait unary_![A]
+  trait unary_+[A]
+  trait N
+  type N1 = -N //Not working
+  type N2 = ~N //Not working
+  type N3 = !N //Not working
+  type N4 = +N //Not working
+}
+```
+
+---
+
+## Proposal
+Add support for prefix types, which is equivalent to the prefix operations for expressions.
+
+```
+PrefixType ::= [`-' | `+' | `~' | `!'] SimpleType
+CompoundType ::= PrefixType
+              |  AnnotType {with AnnotType} [Refinement]
+              |  Refinement
+```
+
+---
+
+## Motivation
+Developers expect terms and types to be expressed the same for mathematical and logical operations.
```

Comment:
Reading just this line I have several questions:
  
1. Can you elaborate on why developers expect that?
1. Perhaps also make a point to make the language more consistent?
1. Can you explain in which use cases or domain are prefix types more likely to be useful?
1. Can you come up with more interesting and *shorter* examples that show the value of prefix types?
1. Can you find other programming languages that support prefix types and tell us about their experience using them?
  
In my opinion, this motivation is not enough to gauge attention from the Committee... motivations are self-contained explanations that research on why the language is lacking without a certain lanaguage feature. They are usually rich in details and examples, and they try to be as objective as possible. For example, saying "developers expect terms and types to be expressed the same for mathematical and logical operations" is not backed up by any empirical or objective data, and it sounds overly subjective.
  
I feel that you can make a much better case for prefix types if you turn your statement around: try to go to the data and analysis first, and then you draw some conclusions from there. Prefix types look like a language feature that has not perceived lots of attention in other languages -- in this case, I think it's important to explain in which scenarios (with concrete examples) would they be useful.

Reply:
I understand what you mean. The problem is justifying this can be just as easily a chicken and egg issue. Nobody uses prefix types because they are not available and the alternative is "good enough". Will a survey passed among the community be acceptable by the committee?

I cannot tell the committee it is not a _niche_ because it is. It will only serve the typelevel community, but it allow us to think _less_ about the language, in general. Currently, the we have to think differently for typelevel operations, because the language made it different. 

Other than the potential for confusion with variance annotation, I don't understand the reluctance include this feature in the language:
* Prefix types are _not_ for doing the same thing but different. They add something that we can't do at all. 
* The implementation is very simple and should not affect compiler in any way.

Reply:
> I understand what you mean. The problem is justifying this can be just as easily a chicken and egg issue. Nobody uses prefix types because they are not available and the alternative is "good enough".
  
If this hasn't been tested before enough, how are we sure that it merits being accepted in the Scala Language Specification? :smile:
  
> Other than the potential for confusion with variance annotation, I don't understand the reluctance to include this feature in the language:
  
Let's see what the Committee thinks. I'll discuss it with Martin too. Thank you for your feedback and updating this document!